torchmultimodal/models/mdetr.py

```diff
@@ -0,0 +1,699 @@
+            pos (Optional[Tensor]): Positional embeddings applied to Q and K
+                self-attention matrices. Default: None
+            query_pos (Optional[Tensor]): Positional embeddings applied to Q
+                cross-attention matrix. Default: None
+    """
+
+    def __init__(
+        self,
+        decoder_layer: nn.Module,
+        num_layers: int,
+        norm: Optional[nn.Module] = None,
+        return_intermediate: bool = False,
+    ):
+        super().__init__()
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self.num_layers = num_layers
+        self.norm = norm
+        self.return_intermediate = return_intermediate
+
+    def forward(
+        self,
+        tgt: Tensor,
+        memory: Tensor,
+        tgt_mask: Optional[Tensor] = None,
+        memory_mask: Optional[Tensor] = None,
+        tgt_key_padding_mask: Optional[Tensor] = None,
+        memory_key_padding_mask: Optional[Tensor] = None,
+        pos: Optional[Tensor] = None,
+        query_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        output = tgt
+
+        intermediate = []
+
+        for layer in self.layers:
+            output = layer(
+                output,
+                memory,
+                tgt_mask=tgt_mask,
+                memory_mask=memory_mask,
+                tgt_key_padding_mask=tgt_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+                pos=pos,
+                query_pos=query_pos,
+            )
+            if self.return_intermediate:
+                intermediate.append(self.norm(output))
+
+        if self.norm is not None:
+            output = self.norm(output)
+            if self.return_intermediate:
+                intermediate.pop()
+                intermediate.append(output)
+
+        if self.return_intermediate:
+            return torch.stack(intermediate)
+
+        return output
+
+
+class TransformerEncoderLayer(nn.Module):
+    """
+    A single layer from a transformer decoder.
+
+    Args:   d_model (int): Number of features in the input.
+            nhead (int): Number of heads in multi-head attention.
+            dim_feedforward (int): Dimension of feedforward network. Default: 2048
+            dropout (float): Dropout value. Default: 0.1.
+            activation (Callable[..., Tensor]): The activation function of the
+                intermediate layer. Default: relu
+            normalize_before (bool): Whether to do PreNorm. Default: False
+
+    Inputs: src (Tensor): The sequence to the encoder layer.
+            src_mask (Optional[Tensor]) The mask for the src sequence. Default: None
+            src_key_padding_mask (Optional[Tensor]): The mask for the src keys per batch.
+                Default: None
+            pos (Optional[Tensor]): Positional embeddings applied to Q and K
+                self-attention matrices. Default: None
+    """
+
+    def __init__(
```

Comment:
justchecking, is this and next class copy paste of nn transformer layer or any logic difference?

Reply:
The main difference is the addition of the `with_pos_embed` method in both classes